internal/mcpproxy/handlers.go

```diff
@@ -588,33 +588,60 @@ func copyProxyHeaders(resp *http.Response, w http.ResponseWriter) {
 
 func (m *MCPProxy) proxyResponseBody(ctx context.Context, s *session, w http.ResponseWriter, resp *http.Response,
 	req *jsonrpc.Request, backend filterapi.MCPBackend,
-) {
+) error {
 	if resp.Header.Get("Content-Type") == "application/json" {
 		body, err := io.ReadAll(resp.Body)
 		if err != nil {
 			m.l.Error("failed to read response body", slog.String("error", err.Error()))
-			return
+			return err
 		}
 		_msg, err := jsonrpc.DecodeMessage(body)
 		if err != nil {
 			m.l.Error("failed to decode JSON-RPC message from response body", slog.String("error", err.Error()))
-			return
+			return err
 		}
 
 		switch msg := _msg.(type) {
 		case *jsonrpc.Request:
 			if err := m.maybeServerToClientRequestModify(ctx, msg, backend.Name); err != nil {
 				m.l.Error("failed to modify server->client request", slog.String("error", err.Error()))
-				return
+				return err
 			}
 			body, _ = jsonrpc.EncodeMessage(msg)
 		case *jsonrpc.Response:
 			if req != nil {
 				if err := m.maybeResponseModify(ctx, req, msg, backend.Name); err != nil {
 					m.l.Error("failed to modify response", slog.String("error", err.Error()))
-					return
+					return err
 				}
 				msg.ID = req.ID
+
+				// Check if this is a JSON-RPC error response
+				if msg.Error != nil {
+					// Return error so metrics are recorded as failure
+					body, _ = jsonrpc.EncodeMessage(msg)
+					w.Header().Set("Content-Length", fmt.Sprintf("%d", len(body)))
+					w.WriteHeader(resp.StatusCode)
+					_, _ = w.Write(body)
+					return fmt.Errorf("%s: %s", string(metrics.MCPErrorInternal), msg.Error.Error())
+				}
+
+				// Check if this is a tools/call response with isError=true
+				if req.Method == "tools/call" && msg.Result != nil {
+					var toolResult mcp.CallToolResult
+					if err := json.Unmarshal(msg.Result, &toolResult); err == nil {
+						if toolResult.IsError {
+							// Tool execution failed, record as failure and return error
+							m.recordResponse(ctx, msg)
+							body, _ = jsonrpc.EncodeMessage(msg)
+							w.Header().Set("Content-Length", fmt.Sprintf("%d", len(body)))
+							w.WriteHeader(resp.StatusCode)
+							_, _ = w.Write(body)
```

Comment:
All this is already done in the default case. can we redo the code to just set the error to return, so we don't duplicate the code?